File: src/api/llmRuntime.ts
const API_BASE = "http://127.0.0.1:8000";

export type LLMRuntimeReadinessState =
  | "unsupported"
  | "missing_runtime"
  | "installing_runtime"
  | "preparing_training_base_snapshot"
  | "runtime_broken"
  | "missing_ollama"
  | "missing_inference_model"
  | "missing_training_base_snapshot"
  | "ready";

export type LLMTrainingMode = "light" | "standard" | "fine";

export type LLMRuntimeModelInfo = {
  name: string;
  sizeBytes: number;
  modifiedAt: string;
  digest: string;
};

export type LLMRuntimeStatus = {
  installed: boolean;
  running: boolean;
  version: string | null;
  minimumVersion: string;
  platform: string;
  models: LLMRuntimeModelInfo[];
  hint: string | null;
  buildVersion?: string | null;
  backendExecutable?: string | null;
  runtimeResourceRoot?: string | null;
  releaseSummaryPath?: string | null;
};

export type LLMRuntimeReadinessAction = {
  id: string;
  label: string;
  description: string;
  recommended: boolean;
};

export type LLMRuntimeInstallProgress = {
  active: boolean;
  percent: number;
  stage: string;
  message: string;
  startedAt: string | null;
  updatedAt: string | null;
  attempt: number;
  errorMessage: string | null;
};

export type LLMRuntimeHardware = {
  gpuModel: string | null;
  vramGB: number;
  driverVersion: string | null;
  cudaVersion: string | null;
  driverCompatibility: "ok" | "incompatible" | "unknown";
  cudaCompatibility: "ok" | "incompatible" | "unknown";
  diskFreeGB: number;
  diskRequiredGB: number;
  source: string;
  supportedModes: LLMTrainingMode[];
};

export type LLMRuntimeReadiness = {
  state: LLMRuntimeReadinessState;
  ready: boolean;
  message: string;
  blockingReason: string | null;
  repairable: boolean;
  actions: LLMRuntimeReadinessAction[];
  installProgress: LLMRuntimeInstallProgress;
  hardware: LLMRuntimeHardware | null;
  checks: Record<string, unknown>;
};

function extractDetail(body: unknown, options?: { status?: number; api?: string }): string {
  if (
    body &&
    typeof body === "object" &&
    "detail" in body &&
    typeof (body as Record<string, unknown>).detail === "string"
  ) {
    const detail = (body as Record<string, unknown>).detail as string;
    if (
      options?.api === "readiness" &&
      options.status === 404 &&
      detail.trim().toLowerCase() === "not found"
    ) {
      return "当前安装包里的后端版本过旧，缺少训练环境接口。请重新运行 Windows 打包脚本并重新安装应用。";
    }
    return detail;
  }
  if (options?.api === "readiness" && options?.status === 404) {
    return "当前安装包里的后端版本过旧，缺少训练环境接口。请重新运行 Windows 打包脚本并重新安装应用。";
  }
  return "操作失败，请稍后重试";
}

export async function fetchLLMRuntime(signal?: AbortSignal): Promise<LLMRuntimeStatus> {
  const resp = await fetch(`${API_BASE}/api/llm/runtime`, { signal });
  const body = await resp.json().catch(() => ({}));
  if (!resp.ok) {
    throw new Error(extractDetail(body, { status: resp.status }));
  }
  return body as LLMRuntimeStatus;
}

export async function fetchLLMRuntimeReadiness(
  options?: {
    mode?: LLMTrainingMode;
    baseModel?: string;
    autoFix?: boolean;
  },
  signal?: AbortSignal,
): Promise<LLMRuntimeReadiness> {
  const query = new URLSearchParams();
  query.set("mode", options?.mode ?? "standard");
  query.set("baseModel", options?.baseModel ?? "qwen2.5:3b");
  query.set("autoFix", options?.autoFix ? "true" : "false");
  const resp = await fetch(`${API_BASE}/api/llm-runtime/readiness?${query.toString()}`, { signal });
  const body = await resp.json().catch(() => ({}));
  if (!resp.ok) {
    throw new Error(extractDetail(body, { status: resp.status, api: "readiness" }));
  }
  return body as LLMRuntimeReadiness;
}

export async function repairLLMRuntime(signal?: AbortSignal): Promise<LLMRuntimeReadiness> {
  const resp = await fetch(`${API_BASE}/api/llm-runtime/repair`, {
    method: "POST",
    signal,
  });
  const body = await resp.json().catch(() => ({}));
  if (!resp.ok) {
    throw new Error(extractDetail(body, { status: resp.status }));
  }
  return body as LLMRuntimeReadiness;
}

export async function openLLMRuntime(signal?: AbortSignal): Promise<void> {
  const resp = await fetch(`${API_BASE}/api/llm/runtime/open`, {
    method: "POST",
    signal,
  });
  if (!resp.ok && resp.status !== 204) {
    const body = await resp.json().catch(() => ({}));
    throw new Error(extractDetail(body, { status: resp.status }));
  }
}

File: backend/app/api/llm_runtime.py
from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, HTTPException, Query, Request, status
from pydantic import BaseModel

from app.services.llm_runtime_manager import (
    LLMRuntimeReadinessState,
)

LLMTrainingMode = Literal["light", "standard", "fine"]

router = APIRouter(prefix="/llm-runtime", tags=["llm-runtime"])


class RuntimeActionPayload(BaseModel):
    id: str
    label: str
    description: str
    recommended: bool


class RuntimeInstallProgressPayload(BaseModel):
    active: bool
    percent: float
    stage: str
    message: str
    startedAt: str | None
    updatedAt: str | None
    attempt: int
    errorMessage: str | None


class RuntimeHardwarePayload(BaseModel):
    gpuModel: str | None
    vramGB: float
    driverVersion: str | None
    cudaVersion: str | None
    driverCompatibility: str
    cudaCompatibility: str
    diskFreeGB: float
    diskRequiredGB: float
    source: str
    supportedModes: list[str]


class LLMRuntimeReadinessPayload(BaseModel):
    state: LLMRuntimeReadinessState
    ready: bool
    message: str
    blockingReason: str | None
    repairable: bool
    actions: list[RuntimeActionPayload]
    installProgress: RuntimeInstallProgressPayload
    hardware: RuntimeHardwarePayload | None
    checks: dict


def _resolve_runtime_manager(request: Request):
    manager = getattr(request.app.state, "llm_runtime_manager", None)
    if manager is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="训练运行时管理器尚未初始化，请稍后重试。",
        )
    return manager


@router.get("/readiness", response_model=LLMRuntimeReadinessPayload)
async def get_llm_runtime_readiness(
    request: Request,
    mode: LLMTrainingMode = Query(default="standard"),
    baseModel: str = Query(default="qwen2.5:3b"),
    autoFix: bool = Query(default=False),
) -> LLMRuntimeReadinessPayload:
    manager = _resolve_runtime_manager(request)
    readiness = await manager.get_readiness(
        mode=mode,
        base_model=baseModel,
        auto_fix=autoFix,
    )
    return LLMRuntimeReadinessPayload(**readiness.to_dict())


@router.post("/repair", response_model=LLMRuntimeReadinessPayload)
async def repair_llm_runtime(request: Request) -> LLMRuntimeReadinessPayload:
    manager = _resolve_runtime_manager(request)
    readiness = await manager.repair_runtime()
    return LLMRuntimeReadinessPayload(**readiness.to_dict())
